bkbit/scripts/gfftranslator.py

```diff
@@ -1,203 +1,682 @@
+        Raises:
+            FileNotFoundError: If the GFF file does not exist.
+
+        Returns:
+            None
+        """
+        gff_file = self.gff_file
+        if self.gff_file.endswith(".gz"):
+            # Decompress the gzip file
+            with gzip.open(self.gff_file, "rb") as f_in:
+                # Create a temporary file to save the decompressed data
+                with tempfile.NamedTemporaryFile(delete=False) as f_out:
+                    # Copy the decompressed data to the temporary file
+                    f_out.write(f_in.read())
+                    gff_file = f_out.name
+
+        if not os.path.isfile(gff_file):
+            raise FileNotFoundError(f"File {gff_file} does not exist.")
+
+        with open(gff_file, "r", encoding="utf-8") as file:
+            curr_line_num = 1
+            progress_bar = tqdm(
+                total=self.__get_line_count(gff_file), desc="Parsing GFF3 File"
+            )
+            for line_raw in file:
+                line_strip = line_raw.strip()
+                if curr_line_num == 1 and not line_strip.startswith("##gff-version 3"):
+                    logger.critical(
+                        'Line %s: ##gff-version 3" missing from the first line.',
+                        curr_line_num,
+                    )
+                elif len(line_strip) == 0:  # blank line
+                    continue
+                elif line_strip.startswith("##"):  # TODO: parse more metadata
+                    pass
+                elif line_strip.startswith("#"):  # TODO: parse more metadata
+                    pass
+                else:  # line may be a feature or unknown
+                    tokens = list(map(str.strip, line_raw.split("\t")))
+                    if len(tokens) != 9:
+                        logger.warning(
+                            "Line %s: Features are expected 9 columns, found %s.",
+                            curr_line_num,
+                            len(tokens),
+                        )
+                    if (
+                        tokens[2] in feature_filter
+                    ):  # only look at rows that have a type that is included in feature_filter
+                        attributes = self.__merge_values(
+                            tuple(a.split("=") for a in tokens[8].split(";"))
+                        )
+                        # TODO: Write cleaner code that calls respective generate function based on the authority automatically
+                        if (
+                            self.genome_annotation.authority
+                            == kbmodel.AuthorityType.ENSEMBL
+                        ):
+                            gene_annotation = self.generate_ensembl_gene_annotation(
+                                attributes, curr_line_num
+                            )
+                            if gene_annotation is not None:
+                                self.gene_annotations[gene_annotation] = gene_annotation
```

Comment:
Did you mean for the key and value of this dict entry to be the same?

Reply:
yes, (based on the document lydia sent) for ensembl genes, entries are considered duplicates if the `id`, `name`, `description`, and `biotype` are the same and those are the attributes the __hash__ function uses. So this is an quick way for me to check for duplicates.

Reply:
That makes sense